walkthrough.ipynb

```diff
@@ -0,0 +1,1950 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "---\n",
+    "- **Documentation and coding style frameworks**\n",
+    "\n",
+    "    - **[PEP 8](https://www.python.org/dev/peps/pep-0008/)** – the official Python style guide for consistent code formatting, naming, indentation, line lengths, etc.\n",
+    "\n",
+    "    - **[PEP 257](https://www.python.org/dev/peps/pep-0257/)** – conventions for writing clear, concise, and structured Python docstrings.\n",
+    "\n",
+    "    - **[NumPy Docstring Standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)** – a widely adopted style for writing detailed, sectioned docstrings, especially in scientific and data projects.\n",
+    "\n",
+    "    - **[Google Python Style Guide](https://google.github.io/styleguide/pyguide.html)** – an alternative coding and docstring style guide, simpler than NumPy style but still structured.\n",
+    "\n",
+    "    - **[reStructuredText (reST)](https://docutils.sourceforge.io/rst.html)** – the markup syntax used in many Python docstrings to format text with headings, lists, links, etc.\n",
+    "\n",
+    "\n",
+    "\n",
+    "---\n",
+    "\n",
+    "- **Notebook environments you may use:**\n",
+    "\n",
+    "    - **[Google Colab](https://colab.research.google.com/)** – free, cloud-hosted Jupyter notebooks with GPU/TPU support.\n",
+    "\n",
+    "    - **[DeepNote](https://deepnote.com/)** – collaborative cloud notebooks with modern UI.\n",
+    "\n",
+    "    - **[JupyterHub](https://iirhub.cloud.e-infra.cz/)** – preconfigured with course resources, allowing you to run pv211-utils notebooks in a set up environment\n",
+    "\n",
+    "    - **[Kaggle Notebooks](https://www.kaggle.com/code)** – cloud notebooks with integrated datasets\n",
+    "\n",
+    "\n",
+    "---\n",
+    "\n",
+    "- **(!)** If you prefer working on your code locally, here are some useful links:\n",
+    "\n",
+    "    - **[JupyterLab](https://jupyter.org/)** – interface for running Jupyter notebooks locally, with tabs, terminals, and some really nice extensions.\n",
+    "\n",
+    "    - **[VS Code Notebooks](https://code.visualstudio.com/docs/datascience/jupyter-notebooks)** – run, edit, and debug notebooks directly in VS Code, with integrated linting and formatting.\n",
+    "\n",
+    "- ...and some frameworks to keep your code clean and work locally:\n",
+    "\n",
+    "    - [flake8](https://flake8.pycqa.org/en/latest/) – checks your code for style issues and simple bugs.\n",
+    "\n",
+    "    - [pydocstyle](https://www.pydocstyle.org/en/stable/) – checks your docstrings follow good conventions.\n",
+    "\n",
+    "    - [black](https://black.readthedocs.io/en/stable/) – auto-formats your code so it always looks neat.\n",
+    "\n",
+    "---"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Cranfield Collection"
```

Comment:
I was thinking maybe it would be nice to also walk the user over the **common concepts** of all the projects: i.e. that there are Documents, Judgements, IRSystem, etc., and how they interact together. Ideally also including the links to the base implementations in the repo.

Reply:
It would also be nice to show how these objects are **used**: you already cover that at the end of the notebook:
```python
results_with_exp = list(retriever_with_exp.search(query))
for doc in results_with_exp:
    print(f"→ Doc {doc.document_id}: {doc.body[:80]}...")
```
...so maybe just copying a small demonstration of this kind to the beginning, before the introduction of specific projects (Cranfield, ...)?